x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java

```diff
@@ -0,0 +1,268 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.index.engine;
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.Fields;
+import org.apache.lucene.index.IndexCommit;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.index.LeafMetaData;
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.NumericDocValues;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.index.StoredFieldVisitor;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.util.Bits;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This special DirectoryReader is used to handle can_match requests against frozen indices.
+ * It' caches all relevant point value data for every point value field ie. min/max packed values etc.
+ * to hold enough information to rewrite a date range query and make a decisions if an index can match or not.
+ * This allows frozen indices to be searched with wildcards in a very efficient way without opening a reader on them.
+ */
+final class RewriteCachingDirectoryReader extends DirectoryReader {
+
+    RewriteCachingDirectoryReader(Directory directory, List<LeafReaderContext> segmentReaders) throws
+        IOException {
+        super(directory, wrap(segmentReaders));
+    }
+
+    private static LeafReader[] wrap(List<LeafReaderContext> readers) throws IOException {
+        LeafReader[] wrapped = new LeafReader[readers.size()];
+        int i = 0;
+        for (LeafReaderContext ctx : readers) {
+            LeafReader wrap = new RewriteCachingLeafReader(ctx.reader());
+            assert wrap != null;
+            wrapped[i++] = wrap;
+        }
+        return wrapped;
+    }
+
+    @Override
+    protected DirectoryReader doOpenIfChanged() {
+        return null;
+    }
+
+    @Override
+    protected DirectoryReader doOpenIfChanged(IndexCommit commit) {
+        return null;
+    }
+
+    @Override
+    protected DirectoryReader doOpenIfChanged(IndexWriter writer, boolean applyAllDeletes) {
+        return null;
+    }
+
+    @Override
+    public long getVersion() {
+        return 0;
+    }
+
+    @Override
+    public boolean isCurrent() {
+        return true;
+    }
+
+    @Override
+    public IndexCommit getIndexCommit() {
+        return null;
+    }
+
+    @Override
+    protected void doClose() {
+
+    }
+
+    @Override
+    public CacheHelper getReaderCacheHelper() {
+        return null;
```

Comment:
I believe all above methods should never be called, should we throw an UOE instead for safety?

Reply:
++